# Credit Card Approval dataset from the UCI Machine Learning Repository

## Explore the dataset

In [5]:
import pandas as pd

In [26]:
# Load dataset
df = pd.read_csv('crx.data',header=None)

In [27]:
df.head(20)

0      1       2  3  4   5  6      7  8  9   10 11 12     13     14 15
0   b  30.83   0.000  u  g   w  v  1.250  t  t   1  f  g  00202      0  +
1   a  58.67   4.460  u  g   q  h  3.040  t  t   6  f  g  00043    560  +
2   a  24.50   0.500  u  g   q  h  1.500  t  f   0  f  g  00280    824  +
3   b  27.83   1.540  u  g   w  v  3.750  t  t   5  t  g  00100      3  +
4   b  20.17   5.625  u  g   w  v  1.710  t  f   0  f  s  00120      0  +
5   b  32.08   4.000  u  g   m  v  2.500  t  f   0  t  g  00360      0  +
6   b  33.17   1.040  u  g   r  h  6.500  t  f   0  t  g  00164  31285  +
7   a  22.92  11.585  u  g  cc  v  0.040  t  f   0  f  g  00080   1349  +
8   b  54.42   0.500  y  p   k  h  3.960  t  f   0  f  g  00180    314  +
9   b  42.50   4.915  y  p   w  v  3.165  t  f   0  t  g  00052   1442  +
10  b  22.08   0.830  u  g   c  h  2.165  f  f   0  t  g  00128      0  +
11  b  29.92   1.835  u  g   c  h  4.335  t  f   0  f  g  00260    200  +
12  a  38.25   6.000  u  g   k  v  1.000  t  f   0  t  g  00000      0  +
13  b  48.08   6.040  u  g   k  v  0.040  f  f   0  f  g  00000   2690  +
14  a  45.83  10.500  u  g   q  v  5.000  t  t   7  t  g  00000      0  +
15  b  36.67   4.415  y  p   k  v  0.250  t  t  10  t  g  00320      0  +
16  b  28.25   0.875  u  g   m  v  0.960  t  t   3  t  g  00396      0  +
17  a  23.25   5.875  u  g   q  v  3.170  t  t  10  f  g  00120    245  +
18  b  21.83   0.250  u  g   d  h  0.665  t  f   0  t  g  00000      0  +
19  a  19.17   8.585  u  g  cc  h  0.750  t  t   7  f  g  00096      0  +

In [28]:
df.tail(20)

0      1       2  3  4   5   6      7  8  9   10 11 12     13   14 15
670  b  47.17   5.835  u  g   w   v  5.500  f  f   0  f  g  00465  150  -
671  b  25.83  12.835  u  g  cc   v  0.500  f  f   0  f  g  00000    2  -
672  a  50.25   0.835  u  g  aa   v  0.500  f  f   0  t  g  00240  117  -
673  ?  29.50   2.000  y  p   e   h  2.000  f  f   0  f  g  00256   17  -
674  a  37.33   2.500  u  g   i   h  0.210  f  f   0  f  g  00260  246  -
675  a  41.58   1.040  u  g  aa   v  0.665  f  f   0  f  g  00240  237  -
676  a  30.58  10.665  u  g   q   h  0.085  f  t  12  t  g  00129    3  -
677  b  19.42   7.250  u  g   m   v  0.040  f  t   1  f  g  00100    1  -
678  a  17.92  10.210  u  g  ff  ff  0.000  f  f   0  f  g  00000   50  -
679  a  20.08   1.250  u  g   c   v  0.000  f  f   0  f  g  00000    0  -
680  b  19.50   0.290  u  g   k   v  0.290  f  f   0  f  g  00280  364  -
681  b  27.83   1.000  y  p   d   h  3.000  f  f   0  f  g  00176  537  -
682  b  17.08   3.290  u  g   i   v  0.335  f  f   0  t  g  00140    2  -
683  b  36.42   0.750  y  p   d   v  0.585  f  f   0  f  g  00240    3  -
684  b  40.58   3.290  u  g   m   v  3.500  f  f   0  t  s  00400    0  -
685  b  21.08  10.085  y  p   e   h  1.250  f  f   0  f  g  00260    0  -
686  a  22.67   0.750  u  g   c   v  2.000  f  t   2  t  g  00200  394  -
687  a  25.25  13.500  y  p  ff  ff  2.000  f  t   1  t  g  00200    1  -
688  b  17.92   0.205  u  g  aa   v  0.040  f  f   0  f  g  00280  750  -
689  b  35.00   3.375  u  g   c   h  8.290  f  f   0  t  g  00000    0  -

In [30]:
# Summary statistics
df_description = df.describe()
print(df_description)

               2           7          10             14
count  690.000000  690.000000  690.00000     690.000000
mean     4.758725    2.223406    2.40000    1017.385507
std      4.978163    3.346513    4.86294    5210.102598
min      0.000000    0.000000    0.00000       0.000000
25%      1.000000    0.165000    0.00000       0.000000
50%      2.750000    1.000000    0.00000       5.000000
75%      7.207500    2.625000    3.00000     395.500000
max     28.000000   28.500000   67.00000  100000.000000


In [31]:
# DataFrame information
df_info = df.info()
print(df_info)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       690 non-null    object 
 1   1       690 non-null    object 
 2   2       690 non-null    float64
 3   3       690 non-null    object 
 4   4       690 non-null    object 
 5   5       690 non-null    object 
 6   6       690 non-null    object 
 7   7       690 non-null    float64
 8   8       690 non-null    object 
 9   9       690 non-null    object 
 10  10      690 non-null    int64  
 11  11      690 non-null    object 
 12  12      690 non-null    object 
 13  13      690 non-null    object 
 14  14      690 non-null    int64  
 15  15      690 non-null    object 
dtypes: float64(2), int64(2), object(12)
memory usage: 86.4+ KB
None


## Handling the missing values

#### The dataset contains float64, int64 and object types. Columns 2, 7, 10 and 14 contain numeric values (of types float64, float64, int64 and int64 respectively) and others contain non-numeric values. 
#### The dataset has missing values "?", now let's replace these missing value question marks with NaN


In [32]:
# Import numpy
import numpy as np

In [33]:
# Replace the '?'s with NaN
df = df.replace('?', np.nan)

In [34]:
#check missing values again
df.tail(20)

0      1       2  3  4   5   6      7  8  9   10 11 12     13   14 15
670    b  47.17   5.835  u  g   w   v  5.500  f  f   0  f  g  00465  150  -
671    b  25.83  12.835  u  g  cc   v  0.500  f  f   0  f  g  00000    2  -
672    a  50.25   0.835  u  g  aa   v  0.500  f  f   0  t  g  00240  117  -
673  NaN  29.50   2.000  y  p   e   h  2.000  f  f   0  f  g  00256   17  -
674    a  37.33   2.500  u  g   i   h  0.210  f  f   0  f  g  00260  246  -
675    a  41.58   1.040  u  g  aa   v  0.665  f  f   0  f  g  00240  237  -
676    a  30.58  10.665  u  g   q   h  0.085  f  t  12  t  g  00129    3  -
677    b  19.42   7.250  u  g   m   v  0.040  f  t   1  f  g  00100    1  -
678    a  17.92  10.210  u  g  ff  ff  0.000  f  f   0  f  g  00000   50  -
679    a  20.08   1.250  u  g   c   v  0.000  f  f   0  f  g  00000    0  -
680    b  19.50   0.290  u  g   k   v  0.290  f  f   0  f  g  00280  364  -
681    b  27.83   1.000  y  p   d   h  3.000  f  f   0  f  g  00176  537  -
682    b  17.08   3.290  u  g   i   v  0.335  f  f   0  t  g  00140    2  -
683    b  36.42   0.750  y  p   d   v  0.585  f  f   0  f  g  00240    3  -
684    b  40.58   3.290  u  g   m   v  3.500  f  f   0  t  s  00400    0  -
685    b  21.08  10.085  y  p   e   h  1.250  f  f   0  f  g  00260    0  -
686    a  22.67   0.750  u  g   c   v  2.000  f  t   2  t  g  00200  394  -
687    a  25.25  13.500  y  p  ff  ff  2.000  f  t   1  t  g  00200    1  -
688    b  17.92   0.205  u  g  aa   v  0.040  f  f   0  f  g  00280  750  -
689    b  35.00   3.375  u  g   c   h  8.290  f  f   0  t  g  00000    0  -

#### Impute the missing values with the mean for numerical variables

In [35]:
df.fillna(df.mean(), inplace=True)

In [36]:
df.isnull().sum()

0     12
1     12
2      0
3      6
4      6
5      9
6      9
7      0
8      0
9      0
10     0
11     0
12     0
13    13
14     0
15     0
dtype: int64

#### Let's replace missing values for columns with objective types

In [37]:
# Iterate over each column of cc_apps
for col in df.columns:
    # Check if the column is of object type
    if df[col].dtypes == 'object':
        # Impute with the most frequent value
        df = df.fillna(df[col].value_counts().index[0])

In [38]:
# Check missing values again
print(df.isnull().sum())

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
dtype: int64


## Preprocessing the data

Convert the non-numeric data into numeric.
Remove unimportant variables.
Split the data into train and test sets.
Scale the feature values to a uniform range.

#### Convert the non-numeric data into numeric

In [39]:
# Import LabelEncoder
from sklearn.preprocessing import LabelEncoder

# Instantiate LabelEncoder
le=LabelEncoder()

# Iterate over all the values of each column and extract their dtypes
for col in df.columns.to_numpy():
    # Compare if the dtype is object
    if df[col].dtypes=='object':
    # Use LabelEncoder to do the numeric transformation
        df[col]=le.fit_transform(df[col])

#### Remove DriversLicense and ZipCode features from the dataset because they are not important for predicting credit card approvals

In [40]:
# Drop the features DriversLicense and ZipCode  and convert the DataFrame to a NumPy array
df = df.drop([11, 13], axis=1)
df = df.to_numpy()

#### Split the data into train and test sets

In [41]:
# Import train_test_split
from sklearn.model_selection import train_test_split

# Segregate features and labels into separate variables
X,y = df[:,0:13] , df[:,13]

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.3,
                                                    random_state=50)

#### Scale the feature values

The credit score of a person is their creditworthiness based on their credit history. The higher this number, the more financially trustworthy a person is considered to be. So, a CreditScore of 1 is the highest since we're rescaling all the values to the range of 0-1.

In [42]:
# Import MinMaxScaler
from sklearn.preprocessing import MinMaxScaler
# Instantiate MinMaxScaler and use it to rescale X_train and X_test
scaler = MinMaxScaler(feature_range=(0, 1))
rescaledX_train = scaler.fit_transform(X_train)
rescaledX_test = scaler.transform(X_test)

## Fitting a logistic regression model to the train set

In [43]:
# Import LogisticRegression
from sklearn.linear_model import LogisticRegression

# Instantiate a LogisticRegression classifier with default parameter values
logreg = LogisticRegression()

# Fit logreg to the train set
logreg.fit(rescaledX_train,y_train)

LogisticRegression()

#### Predict and evaluate performance

In [44]:
# Import confusion_matrix
from sklearn.metrics import confusion_matrix

# Use logreg to predict instances from the test set and store it
y_pred = logreg.predict(rescaledX_test)

# Get the accuracy score of logreg model and print it
print("Accuracy of logistic regression: ", logreg.score(rescaledX_test,y_test))

# Print the confusion matrix of the logreg model
confusion_matrix(y_test,y_pred)

Accuracy of logistic regression:  0.8840579710144928


array([[85,  9],
       [15, 98]])

The model was pretty good. It was able to yield an accuracy score of almost 88%

For the confusion matrix, the first element of the of the first row of the confusion matrix denotes the true negatives meaning the number of negative instances (denied applications) predicted by the model correctly. And the last element of the second row of the confusion matrix denotes the true positives meaning the number of positive instances (approved applications) predicted by the model correctly.